core/src/sumeragi/main_loop.rs

```diff
@@ -699,113 +776,86 @@ pub fn run<F>(
-                            // in order to preserve old behaviour. This should be changed.
-                            // Tracking issue : https://github.com/hyperledger/iroha/issues/2635
-                            let block = match block.revalidate(
-                                &sumeragi.transaction_validator,
-                                &state.wsv,
-                                &state.latest_block_hash,
-                                state.latest_block_height,
-                            ) {
-                                Ok(block) => block,
-                                Err(err) => {
-                                    warn!(%err);
-                                    continue;
-                                }
-                            };
-                            for event in Vec::<Event>::from(&block) {
-                                trace!(?event);
-                                sumeragi.events_sender.send(event).unwrap_or(0);
+                        continue;
+                    }
+
+                    let block = {
+                        let span = span!(
+                            Level::TRACE,
+                            "Sumeragi Observing Peer is revalidating the block."
+                        );
+                        let _enter = span.enter();
+                        match block.revalidate(
+                            &sumeragi.transaction_validator,
+                            &state.wsv,
+                            &state.latest_block_hash,
+                            state.latest_block_height,
+                        ) {
+                            Ok(block) => block,
+                            Err(err) => {
+                                warn!(%err);
+                                continue;
                             }
-                            commit_block(sumeragi, block, &mut state);
                         }
+                    };
+
+                    for event in Vec::<Event>::from(&block) {
+                        trace!(?event);
+                        sumeragi.events_sender.send(event).unwrap_or(0);
                     }
-                    _ => {
-                        trace!("Observing peer not handling message {:?}", incoming_message);
-                    }
+
+                    let block_clone = block.clone();
+                    let key_pair_clone = sumeragi.key_pair.clone();
+                    let signed_block = block_clone
+                        .sign(key_pair_clone)
+                        .expect("maybe we should handle this error");
+
+                    let voting_block = VotingBlock::new(signed_block.clone());
+                    voting_block_option = Some(voting_block);
                 }
-            } else {
-                should_sleep = true;
+                Some(msg) => {
+                    trace!(?msg, "Observing peer not handling message");
```

Comment:
```suggestion
                    trace!(?msg, "Sumeragi Observing Peer is not handling the message");
```

though it's not entirely clear to me what exactly this means and entails

Reply:
> though it's not entirely clear to me what exactly this means and entails

 This is needed for debugging purposes to detect if we forget to handle message we suppose to handle.